File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/outsleeping/repository/OutSleepingRepository.java
package b1nd.dodam.domain.rds.outsleeping.repository;

import b1nd.dodam.domain.rds.member.entity.Student;
import b1nd.dodam.domain.rds.outsleeping.entity.OutSleeping;
import b1nd.dodam.domain.rds.outsleeping.exception.OutSleepingNotFoundException;
import org.springframework.data.jpa.repository.EntityGraph;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDate;
import java.util.List;

public interface OutSleepingRepository extends JpaRepository<OutSleeping, Long> {

    default OutSleeping getById(Long id) {
        return findById(id)
                .orElseThrow(OutSleepingNotFoundException::new);
    }

    @EntityGraph(attributePaths = {"student.member"})
    @Query("select o from OutSleeping o where :date between o.startAt and o.endAt " +
            "and o.endAt > :date " +
            "and o.status = 'ALLOWED' " +
            "order by o.student.grade, o.student.room, o.student.number")
    List<OutSleeping> findByDate(@Param("date") LocalDate date);

    @EntityGraph(attributePaths = {"student.member"})
    List<OutSleeping> findByEndAtGreaterThanEqual(LocalDate endAt);

    @EntityGraph(attributePaths = {"student.member"})
    List<OutSleeping> findByStudentAndEndAtGreaterThanEqual(Student student, LocalDate now);

}

File: services/service-outsleeping/src/main/kotlin/com/b1nd/dodamdodam/outsleeping/domain/outsleeping/repository/OutSleepingRepository.kt
package com.b1nd.dodamdodam.outsleeping.domain.outsleeping.repository

import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.entity.OutSleepingEntity
import com.b1nd.dodamdodam.outsleeping.domain.outsleeping.enumeration.OutSleepingStatusType
import org.springframework.data.jpa.repository.JpaRepository
import java.time.LocalDate
import java.util.UUID

interface OutSleepingRepository : JpaRepository<OutSleepingEntity, Long>, OutSleepingRepositoryCustom {

    // 날짜별 조회: startAt <= date <= endAt 인 외박 전체
    fun findByStartAtLessThanEqualAndEndAtGreaterThanEqual(
        endDate: LocalDate,
        startDate: LocalDate
    ): List<OutSleepingEntity>

    // 내 외박 조회
    fun findByStudentId(studentId: UUID): List<OutSleepingEntity>

    // 유효한 외박 조회: ALLOWED + 오늘이 startAt~endAt 범위 안
    fun findByStatusAndStartAtLessThanEqualAndEndAtGreaterThanEqual(
        status: OutSleepingStatusType,
        endDate: LocalDate,
        startDate: LocalDate
    ): List<OutSleepingEntity>
}
